[GEN][ZH] Fix Retail Mismatch in Release Build when Logging is enabled

helmutbuhler opened the pull request:
This PR disables some logging in Releasebuild when Logging is enabled that causes a Mismatch. It'll be nice to have logging work in release builds to debug mismatch issues.

With this fix applied, you can load a retail replay with AI players and play it back in a Release build with logging without mismatches.

helmutbuhler commented:
Yes, sin and cos are internally implemented with doubles. If the logging is not there, the optimizer removes the casts to float and uses the doubles in the final calculation. If the logging is there, it doesn't remove the cast because the float is logged, and then also uses the float in the gamelogic.
This happens in quite a few places, but luckily only this log call seems to cause a gamelogic change.

xezon commented:
This is outrageous lol. Nice. Can this perhaps also be fixed by moving the logging a few lines?

helmutbuhler commented:
I highly doubt it. Also keep in mind that some other log calls are already guarded by `#if defined(_DEBUG) || defined(_INTERNAL)`. One of them in the same file. The original devs were possibly aware of this issue and just missed this one Log call.

roossienb commented:
Does this mean that after the fix, the debug version of the game mismatches against the release version of the game? That could hamper testing efforts. Or does it mismatch already?

helmutbuhler commented:
Yes, VC6 Debug and Internal Version immediately mismatch against retail. Fixing that would be very difficult. I initially planned to do that and published that work here: https://github.com/TheSuperHackers/GeneralsGameCode/pull/540
But it's so much work, it's not really feaseble.

xezon commented:
I tested this change with replay of https://github.com/roossienb/GeneralsWiki/blob/e52ca6618f90cef9ece00cbb7abacc293f580b12/SourceCode/Debug/files/Zerohour-skirmish.zip and it mismatched with vs6 release build + RTS_DEBUG_LOGGING=ON + RTS_DEBUG_CRASHING=ON. Did I do something wrong?

helmutbuhler commented:
> I tested this change with replay of https://github.com/roossienb/GeneralsWiki/blob/e52ca6618f90cef9ece00cbb7abacc293f580b12/SourceCode/Debug/files/Zerohour-skirmish.zip and it mismatched with vs6 release build + RTS_DEBUG_LOGGING=ON + RTS_DEBUG_CRASHING=ON. Did I do something wrong?

Thanks for testing! You are right, looks like there is something else that causes mismatches. I'll try to find that issue as well. It's a bummer though because it's really annoying to find these things. Took me almost an entire day to find this one in this PR.

Btw I used the attached replay for this PR, in case you wanna check. Someone on Discord provided it.
[tansooo.zip](https://github.com/user-attachments/files/19917479/tansooo.zip)
[TANSO AI 1v1 GLA Mirror.zip](https://github.com/user-attachments/files/19917484/TANSO.AI.1v1.GLA.Mirror.zip)

helmutbuhler commented:
> I tested this change with replay of https://github.com/roossienb/GeneralsWiki/blob/e52ca6618f90cef9ece00cbb7abacc293f580b12/SourceCode/Debug/files/Zerohour-skirmish.zip and it mismatched with vs6 release build + RTS_DEBUG_LOGGING=ON + RTS_DEBUG_CRASHING=ON. Did I do something wrong?

Just pushed a fix for it. This time it's likely different code due to different inlining.

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/Common/StateMachine.cpp:
```diff
@@ -518,15 +518,19 @@ State *StateMachine::internalGetState( StateID id )
+#endif
 			throw ERROR_BAD_ARG;
-		} else {
-			DEBUG_LOG(("Located default state to recover.\n"));
```
I removed this else block because it's redundant

xezon replied:
Move this and above change in another pull request as well. It does not belong here.

helmutbuhler commented on Generals/Code/GameEngine/Source/Common/StateMachine.cpp:
```diff
@@ -499,8 +499,20 @@ State *StateMachine::internalGetState( StateID id )
 	{
-		DEBUG_CRASH(( "StateMachine::internalGetState(): Invalid state" ));
-		throw ERROR_BAD_ARG;
+		// TheSuperHackers @info helmutbuhler 26/04/2025
```
I replaced this with the ZH version

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/StateMachine.cpp:
```diff
@@ -518,15 +518,19 @@ State *StateMachine::internalGetState( StateID id )
 		DEBUG_CRASH( ("StateMachine::internalGetState(): Invalid state for object %s using state %d", m_owner->getTemplate()->getName().str(), id) );
-		DEBUG_LOG(("Transisioning to state #d\n", (Int)id));
+#if defined(_DEBUG) || defined(_INTERNAL)
+		DEBUG_LOG(("Transitioning to state #d\n", (Int)id));
```
%d

xezon replied:
Can you fix this while at it?

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/StateMachine.cpp:
```diff
@@ -518,15 +518,19 @@ State *StateMachine::internalGetState( StateID id )
 	{
+		// TheSuperHackers @info helmutbuhler 26/04/2025
+		// These logs cause mismatch on Release when logging is enabled.
 		DEBUG_CRASH( ("StateMachine::internalGetState(): Invalid state for object %s using state %d", m_owner->getTemplate()->getName().str(), id) );
```
Should this be included?

xezon replied:
DEBUG_CRASH is enabled when using the RTS_DEBUG_CRASHING cmake option, which is useful to get the assert message for Replay mismatch.

xezon reviewed (commented):
How do you come to the conclusion that these logs cause mismatch in particular?

xezon commented:
> https://github.com/user-attachments/files/19917484/TANSO.AI.1v1.GLA.Mirror.zip

This replay cannot be used because the map is unknown.

xezon commented:
I tested this updated change against https://github.com/roossienb/GeneralsWiki/blob/e52ca6618f90cef9ece00cbb7abacc293f580b12/SourceCode/Debug/files/Zerohour-skirmish.zip and it still mismatches.

How do you determine what logging would be mismatching?

roossienb commented:
> I tested this updated change against https://github.com/roossienb/GeneralsWiki/blob/e52ca6618f90cef9ece00cbb7abacc293f580b12/SourceCode/Debug/files/Zerohour-skirmish.zip and it still mismatches.
> 
> Edit: The above replay also goes out of sync without Release logging. Perhaps that replay is bad?
> 
> How do you determine what logging would be mismatching?

I tested the replay against the latest MAIN build and have no mismatches.

helmutbuhler commented:
> How do you determine what logging would be mismatching?

I basically do a binary search over the cpp files. I disable the DEBUG_LOG macro in individual files and pin point to the file, then to the log call.

xezon commented:
> I basically do a binary search over the cpp files. I disable the DEBUG_LOG macro in individual files and pin point to the file, then to the log call.

Uff that is tough. I was hoping that the CRC logging would help pinpointing it.

helmutbuhler commented:
> > I basically do a binary search over the cpp files. I disable the DEBUG_LOG macro in individual files and pin point to the file, then to the log call.
> 
> Uff that is tough. I was hoping that the CRC logging would help pinpointing it.

CRC logging can also cause mismatches ;) The problem isn't that something is written to disk, but that the optimizer produces different machinecode.

xezon commented:
I am unable to confirm that this change fixes mismatching when logging is enabled.

helmutbuhler commented:
Alright, I did some extensive testing and found a few more Log calls that caused mismatches. I put 4 more #ifdefs around some log calls. I also fixed a runtime error when the version string that is stored in the replay has some russian characters.

With these changes, I could successfully simulate 1400 replays that were uploaded on the Gentool server without mismatches, so I feel pretty confident that logging doesn't cause any false mismatches anymore. But of course there is no 100% guarantee.
I also made sure that DEBUG_LOGGING and DEBUG_CRASHING works, initially I only enabled DEBUG_LOGGING.

I also found out that the log in StateMachine::internalGetState only causes mismatches when the supplyTruck variable is kept uninitialized. So I also checked if initializing that variable causes any mismatches, and it doesn't. So my suggestion is to initialize that variable (I'll make a PR) and we can keep the logging here in Release mode (I took the #ifdefs out)

I attach here 3 replays that will cause mismatches in Release with DEBUG_LOGGING and DEBUG_CRASHING defined and this PR not applied. Each of the 3 files where I added #ifdefs fixes one of the replays.
[release_logging_desyncs.zip](https://github.com/user-attachments/files/20065123/release_logging_desyncs.zip)

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/Recorder.cpp:
```diff
@@ -1118,7 +1118,10 @@ Bool RecorderClass::playbackFile(AsciiString filename)
-			tempStr.format("   Version [%ls] vs [%ls]\n", TheVersion->getUnicodeVersion().str(), header.versionString.str());
+			// AsciiString::format with %ls can make internal _vsnprintf call fail on some strings. We convert using translate here. (TheSuperHackers @fix helmutbuhler 05/05/2025)
+			UnicodeString tempStrWide;
+			tempStrWide.format(L"   Version [%s] vs [%s]\n", TheVersion->getUnicodeVersion().str(), header.versionString.str());
```
This fix needs to go into a separate pull request. Also, I am left a bit puzzled by the reason for it. Why does the refactor work? And also there are more places that do this, so all of them need checking?

helmutbuhler replied:
Yeah, the whole Unicode conversion is a big mess. Won't be easy to clean up. But for now I only need this for replay simulation. I'll put it in another PR...

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/Recorder.cpp:
```diff
@@ -1118,7 +1118,10 @@ Bool RecorderClass::playbackFile(AsciiString filename)
 		if (versionStringDiff)
 		{
-			tempStr.format("   Version [%ls] vs [%ls]\n", TheVersion->getUnicodeVersion().str(), header.versionString.str());
+			// AsciiString::format with %ls can make internal _vsnprintf call fail on some strings. We convert using translate here. (TheSuperHackers @fix helmutbuhler 05/05/2025)
```
Comment format starts with "TheSuperHackers ..."

helmutbuhler commented:
So can this be merged now?

xezon commented:
I wanted to take a look at the assembler to understand what causes this.

helmutbuhler commented:
> I wanted to take a look at the assembler to understand what causes this.

It's likely different inlining, which results in different constant propagation, which results in different fast-math-optimization

Caball009 commented:
> Yes, sin and cos are internally implemented with doubles. If the logging is not there, the optimizer removes the casts to float and uses the doubles in the final calculation. If the logging is there, it doesn't remove the cast because the float is logged, and then also uses the float in the gamelogic. This happens in quite a few places, but luckily only this log call seems to cause a gamelogic change.

Floating point arithmetic is complicated: https://isocpp.org/wiki/faq/newbie#floating-point-arith2

helmutbuhler commented:
> Floating point arithmetic is complicated: https://isocpp.org/wiki/faq/newbie#floating-point-arith2

My understanding is that those points in the faq only apply when the fast math option in the compiler is set. When it's set to precise, all float operations are well defined across compilers and cpus.

helmutbuhler commented:
> Only GCC elides the repeated calls to std::cos.

So? I don't understand your conclusion here.

Caball009 commented:
> When it's set to precise, all float operations are well defined across compilers and cpus.

I showed that GCC behaves differently from Clang and MSVC if fast math isn't enabled. How's that well-defined across compilers?

OmniBlade commented:
GCC and Clang as well as modern MSVC will use SSE instructions and are more strict about correctly rounding when requested to do so with casts. x87 which VC6 uses always expands to double precision in its maths, no matter what, so would be very difficult to match on any other instruction set and behaves oddly if code changes alter if a value is kept in register or spills to variable.

Would be interesting if log/no log builds on MSVC 2022 desync with each other before this fix.

helmutbuhler commented:
> > When it's set to precise, all float operations are well defined across compilers and cpus.
> 
> It shows that GCC behaves differently from Clang and MSVC if fast math isn't enabled. How's that well-defined across compilers?

You just showed, that the assembly is a bit different. The observable behavior of the produced program is the same in all compilers, isn't it?

xezon commented on Generals/Code/GameEngine/Source/GameLogic/AI/AIStates.cpp:
```diff
@@ -4028,8 +4028,11 @@ StateReturnType AIFollowWaypointPathState::update()
 		if (getAdjustsDestination() && ai->isDoingGroundMovement()) {
 			if (!TheAI->pathfinder()->adjustDestination(obj, ai->getLocomotorSet(), &m_goalPosition)) {
 				if (m_currentWaypoint) {
+					// This log causes mismatch on Release (TheSuperHackers @info helmutbuhler 05/05/2025)
```
Better move this behind the #ifdef like the other places. And fix the comment format according to our comment style: `TheSuperHackers @info Log causes mismatch in Release`. Several times in this change.

xezon commented on GeneralsMD/Code/GameEngine/Source/GameLogic/AI/AISkirmishPlayer.cpp:
```diff
@@ -681,12 +681,16 @@ void AISkirmishPlayer::buildAIBaseDefenseStructure(const AsciiString &thingName,
-	  DEBUG_LOG(("buildAIBaseDefenseStructure -- Angle is %f sin %f, cos %f \n", 180*angle/PI, s, c));
+#if defined(RTS_DEBUG) || defined(RTS_INTERNAL)
+		// TheSuperHackers @info helmutbuhler 21/04/2025
+		// This log causes mismatch on Release when logging is enabled.
```
These multiline comments could be simplified to `// TheSuperHackers @info Log causes mismatch in Release` for compactness.

xezon commented:
Comment format needs fixing because it is inconsistent with all our other comments.
